File: packages/share/ios/Classes/SharePlugin.m
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "SharePlugin.h"

@interface NSError (FlutterError)
@property(readonly, nonatomic) FlutterError *flutterError;
@end

@implementation NSError (FlutterError)
- (FlutterError *)flutterError {
  return [FlutterError errorWithCode:[NSString stringWithFormat:@"Error %d", (int)self.code]
                             message:self.domain
                             details:self.localizedDescription];
}
@end

static NSString *const PLATFORM_CHANNEL = @"plugins.flutter.io/share";

@implementation FLTSharePlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FlutterMethodChannel *shareChannel =
      [FlutterMethodChannel methodChannelWithName:PLATFORM_CHANNEL
                                  binaryMessenger:registrar.messenger];

  [shareChannel setMethodCallHandler:^(FlutterMethodCall *call, FlutterResult result) {
    NSDictionary *arguments = [call arguments];
    if ([@"openShareDialog" isEqualToString:call.method]) {
      NSString *shareText = arguments[@"text"];

      if (shareText.length == 0) {
        result(
            [FlutterError errorWithCode:@"error" message:@"Non-empty text expected" details:nil]);
        return;
      }

      NSNumber *originX = arguments[@"originX"];
      NSNumber *originY = arguments[@"originY"];
      NSNumber *originWidth = arguments[@"originWidth"];
      NSNumber *originHeight = arguments[@"originHeight"];

      CGRect originRect;
      if (originX != nil && originY != nil && originWidth != nil && originHeight != nil) {
        originRect = CGRectMake([originX doubleValue], [originY doubleValue],
                                [originWidth doubleValue], [originHeight doubleValue]);
      }

      [self openShareDialog:shareText
             withController:[UIApplication sharedApplication].keyWindow.rootViewController
                   atSource:originRect];
      result(nil);
    } else if ([@"listKnownAvailableSharingApps" isEqualToString:call.method]) {
      [self listKnownAvailableSharingApps:result];
    } else if ([@"shareWithTwitter" isEqualToString:call.method]) {
      [self shareWithTwitter:result text:arguments[@"text"]];
    } else if ([@"shareWithTelegram" isEqualToString:call.method]) {
      [self shareWithTelegram:result text:arguments[@"text"]];
    } else if ([@"shareWithWhatsApp" isEqualToString:call.method]) {
      [self shareWithWhatsApp:result text:arguments[@"text"]];
    } else if ([@"shareWithEmail" isEqualToString:call.method]) {
      [self shareWithEmail:result
                      text:arguments[@"text"]
                 recipient:arguments[@"recipient"]
                   subject:arguments[@"subject"]];
    } else {
      result(FlutterMethodNotImplemented);
    }
  }];
}

+ (void)openShareDialog:(id)sharedItems
         withController:(UIViewController *)controller
               atSource:(CGRect)origin {
  UIActivityViewController *activityViewController =
      [[UIActivityViewController alloc] initWithActivityItems:@[ sharedItems ]
                                        applicationActivities:nil];
  activityViewController.popoverPresentationController.sourceView = controller.view;
  if (!CGRectIsEmpty(origin)) {
    activityViewController.popoverPresentationController.sourceRect = origin;
  }
  [controller presentViewController:activityViewController animated:YES completion:nil];
}

+ (void)listKnownAvailableSharingApps:(FlutterResult)result {
  UIApplication *application = [UIApplication sharedApplication];
  NSMutableArray *availableApps = [NSMutableArray new];
  if ([application canOpenURL:[NSURL URLWithString:@"twitter://post?message="]]) {
    [availableApps addObject:@"twitter"];
  }
  if ([application canOpenURL:[NSURL URLWithString:@"tg://msg?text="]]) {
    [availableApps addObject:@"telegram"];
  }
  if ([application canOpenURL:[NSURL URLWithString:@"whatsapp://send?text="]]) {
    [availableApps addObject:@"whatsapp"];
  }
  if ([application canOpenURL:[NSURL URLWithString:@"mailto:foo@bar?body="]]) {
    [availableApps addObject:@"email"];
  }
  result(availableApps);
}

+ (void)shareWithTwitter:(FlutterResult)result text:(NSString *)text {
  NSString *urlString = [[NSString stringWithFormat:@"twitter://post?message=%@", text]
      stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
  [self openUrl:result url:[NSURL URLWithString:urlString]];
}

+ (void)shareWithTelegram:(FlutterResult)result text:(NSString *)text {
  NSString *urlString = [[NSString stringWithFormat:@"tg://msg?text=%@", text]
      stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
  [self openUrl:result url:[NSURL URLWithString:urlString]];
}

+ (void)shareWithWhatsApp:(FlutterResult)result text:(NSString *)text {
  NSString *urlString = [[NSString stringWithFormat:@"whatsapp://send?text=%@", text]
      stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
  [self openUrl:result url:[NSURL URLWithString:urlString]];
}

+ (void)shareWithEmail:(FlutterResult)result
                  text:(NSString *)text
             recipient:(NSString *)recipient
               subject:(NSString *)subject {
  NSString *urlString =
      [[NSString stringWithFormat:@"mailto:%@?subject=%@&body=%@", recipient, subject, text]
          stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
  [self openUrl:result url:[NSURL URLWithString:urlString]];
}

+ (void)openUrl:(FlutterResult)result url:(NSURL *)url {
  UIApplication *application = [UIApplication sharedApplication];
  if ([application canOpenURL:url]) {
    [application openURL:url];
    result(nil);
  } else {
    result([FlutterError
        errorWithCode:@"shareError"
              message:[NSString stringWithFormat:@"Couldn't open app with url = %@", url]
              details:nil]);
  }
}
@end

File: packages/share/android/src/main/java/io/flutter/plugins/share/SharePlugin.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.share;

import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/** Plugin method host for presenting a share sheet via Intent */
public class SharePlugin implements MethodChannel.MethodCallHandler {

  private static final String CHANNEL = "plugins.flutter.io/share";

  /// The URL is used by [listKnownAvailableSharingApps] to test if the app is available on the device.
  private static final Map<String, String> KNOWN_SHARING_APPS =
      new HashMap<String, String>() {
        {
          put("twitter", "twitter://post?message=");
          put("telegram", "tg://msg?text=");
          put("whatsapp", "whatsapp://send?text=");
          put("email", "mailto:foo@bar?body=");
        }
      };

  public static void registerWith(Registrar registrar) {
    MethodChannel channel = new MethodChannel(registrar.messenger(), CHANNEL);
    SharePlugin instance = new SharePlugin(registrar);
    channel.setMethodCallHandler(instance);
  }

  private final Registrar mRegistrar;

  private SharePlugin(Registrar registrar) {
    this.mRegistrar = registrar;
  }

  @Override
  public void onMethodCall(MethodCall call, MethodChannel.Result result) {
    if (call.method.equals("openShareDialog")) {
      if (!(call.arguments instanceof Map)) {
        throw new IllegalArgumentException("Map argument expected");
      }
      // Android does not support showing the share sheet at a particular point on screen.
      openShareDialog((String) call.argument("text"));
      result.success(null);
    } else if (call.method.equals("listKnownAvailableSharingApps")) {
      listKnownAvailableSharingApps(result);
    } else if (call.method.equals("shareWithTwitter")) {
      shareWithTwitter(result, (String) call.argument("text"));
    } else if (call.method.equals("shareWithTelegram")) {
      shareWithTelegram(result, (String) call.argument("text"));
    } else if (call.method.equals("shareWithWhatsApp")) {
      shareWithWhatsApp(result, (String) call.argument("text"));
    } else if (call.method.equals("shareWithEmail")) {
      shareWithEmail(
          result,
          (String) call.argument("text"),
          (String) call.argument("recipient"),
          (String) call.argument("subject"));
    } else {
      result.notImplemented();
    }
  }

  private void openShareDialog(String text) {
    if (text == null || text.isEmpty()) {
      throw new IllegalArgumentException("Non-empty text expected");
    }

    Intent shareIntent = new Intent();
    shareIntent.setAction(Intent.ACTION_SEND);
    shareIntent.putExtra(Intent.EXTRA_TEXT, text);
    shareIntent.setType("text/plain");
    Intent chooserIntent = Intent.createChooser(shareIntent, null /* dialog title optional */);
    if (mRegistrar.activity() != null) {
      mRegistrar.activity().startActivity(chooserIntent);
    } else {
      chooserIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
      mRegistrar.context().startActivity(chooserIntent);
    }
  }

  private void listKnownAvailableSharingApps(MethodChannel.Result result) {
    PackageManager packageManager = mRegistrar.activity().getPackageManager();
    ArrayList<String> availableApps = new ArrayList<>();
    for (Map.Entry<String, String> entry : KNOWN_SHARING_APPS.entrySet()) {
      Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(entry.getValue()));
      if (intent.resolveActivity(packageManager) != null) {
        availableApps.add(entry.getKey());
      }
    }
    result.success(availableApps);
  }

  private void shareWithTwitter(MethodChannel.Result result, String text) {
    openUrl(result, Uri.parse("twitter://post?message=" + text));
  }

  private void shareWithTelegram(MethodChannel.Result result, String text) {
    openUrl(result, Uri.parse("tg://msg?text=" + text));
  }

  private void shareWithWhatsApp(MethodChannel.Result result, String text) {
    openUrl(result, Uri.parse("whatsapp://send?text=" + text));
  }

  private void shareWithEmail(
      MethodChannel.Result result, String text, String recipient, String subject) {
    openUrl(result, Uri.parse("mailto:" + recipient + "?subject=" + subject + "&body=" + text));
  }

  private void openUrl(MethodChannel.Result result, Uri uri) {
    try {
      mRegistrar.activity().startActivity(new Intent(Intent.ACTION_VIEW, uri));
    } catch (ActivityNotFoundException e) {
      result.error("shareError", e.getMessage(), null);
      return;
    }
    result.success(null);
  }
}
